Sources/SPMBuildCore/PluginInvocation.swift

```diff
@@ -126,30 +116,47 @@ extension PackageGraph {
                     }
                 }
                 
-                // Generate commands from the plugin output.
+                // Generate commands from the plugin output.  This is where we translate from the transport JSON to our internal form.
                 let commands: [PluginInvocationResult.Command] = pluginOutput.commands.map { cmd in
                     let displayName = cmd.displayName
                     let executable = cmd.executable
                     let arguments = cmd.arguments
-                    let workingDir = cmd.workingDirectory.map{ AbsolutePath($0) }
                     let environment = cmd.environment
+                    let workingDir = cmd.workingDirectory.map{ AbsolutePath($0) }
                     switch extTarget.capability {
                     case .prebuild:
-                        let derivedSourceDirPaths = cmd.derivedSourcePaths.map{ AbsolutePath($0) }
-                        return .prebuildCommand(displayName: displayName, executable: executable, arguments: arguments, workingDir: workingDir, environment: environment, derivedSourceDirPaths: derivedSourceDirPaths)
+                        return .prebuildCommand(
+                            displayName: displayName,
+                            executable: executable,
+                            arguments: arguments,
+                            environment: environment ?? [:],
```

Comment:
nil: should environment be optional?

Reply:
Possibly.  In general, it seems that collections are represented as non-optional but with empty defaults in our code base.  The internal representation of this actually was optional in the WIP implementation but it seems annoying to have to unwrap at each point of use (I am assuming that empty collections do not incur extra heap allocations etc so that it's mainly a stylistic choice).  Do you or @neonichu have strong view on this?  If so then the arrays be optional too?

Reply:
personally I prefer to use optional where the data may be missing, and "pipe" it through the call stack as optional all the way so that the unwrapping only happen in the very end

Reply:
There are definitely cases in which there are semantic differences between nil and empty set, but in this case (the environment variables to set in the subprocess) there doesn't seem to be a useful distinction between nil and empty.